### PR TITLE
[GOBBLIN-1672] Refactor metrics from DagManager into its own class, add metrics per …

### PR DESCRIPTION
**Pull request by Will-Lo**
…executor for SUCCESS, FAILED, SLA_EXCEEDED, START_SLA_EXCEEDED

Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [x] My PR addresses the following [Gobblin JIRA](https://issues.apache.org/jira/browse/GOBBLIN/) issues and references them in the PR title. For example, "[GOBBLIN-XXX] My Gobblin PR"
    - https://issues.apache.org/jira/browse/GOBBLIN-1672

### Description
- [x] Here are some details about my PR, including screenshots (if applicable):
- Add more metrics related to flow management. Since a good chunk of the code in the DagManager was related to metrics emission, I tried to contain it within a class.
- Adds the following metrics:
1. Successes per executor
2. Fails per executor
3. SLA Exceeded per executor
4. SLA Exceeded per flowgroup
5. Total SLA Exceeded
6. Jobs sent to spec executors
7.  Start SLA exceeded per executor

### Tests
- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    5. Subject uses the imperative mood ("add", not "adding")
    6. Body wraps at 72 characters
    7. Body explains "what" and "why", not "how"

### CODE REVIEW

File: gobblin-service/src/test/java/org/apache/gobblin/service/modules/orchestration/DagManagerTest.java [hunk 1]
```diff
@@ -1053,6 +1062,150 @@ public void testEmitFlowMetricOnlyIfNotAdhoc() throws URISyntaxException, IOExce
     Assert.assertEquals(metricContext.getParent().get().getGauges().get(flowStateGaugeName1).getValue(), DagManager.FlowState.SUCCESSFUL.value);
   }
 
+  @Test (dependsOnMethods = "testEmitFlowMetricOnlyIfNotAdhoc")
+  public void testJobSlaKilledMetrics() throws URISyntaxException, IOException {
+    long flowExecutionId = System.currentTimeMillis() - 10 * 60 * 1000;
```
Comment (arjun4084346) [inline gobblin-service/src/test/java/org/apache/gobblin/service/modules/orchestration/DagManagerTest.java:1067]: *nit 
I use Duration.ofMinutes(10).toMillis() to keep it more readable

File: gobblin-service/src/test/java/org/apache/gobblin/service/modules/orchestration/DagManagerTest.java [hunk 2]
```diff
@@ -779,15 +785,18 @@ public void testJobKilledSLAMetricsArePerExecutor() throws URISyntaxException, I
         thenReturn(jobStatusIterator2).
         thenReturn(jobStatusIterator3);
 
-    // Run the thread once. All 3 jobs should be emitted an sla exceeded event
+    // Run the thread once. All 3 jobs should be emitted a sla exceeded event
```
Comment (arjun4084346) [inline gobblin-service/src/test/java/org/apache/gobblin/service/modules/orchestration/DagManagerTest.java:788]: "an" might be right, https://www.reddit.com/r/ENGLISH/comments/433x3y/do_you_use_a_or_an_before_the_acronym_sla_sla_is/
I will let you decide on this :D

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/DagManagerMetrics.java [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+  // These metrics need to be invoked differently to account for automated retries and multihop scenarios.
+  private final Map<String, ContextAwareMeter> executorSuccessMeters = Maps.newConcurrentMap();
+  private final Map<String, ContextAwareMeter> executorFailureMeters = Maps.newConcurrentMap();
+  private final Map<String, ContextAwareMeter> executorStartSlaExceededMeters = Maps.newConcurrentMap();
+  private final Map<String, ContextAwareMeter> executorSlaExceededMeters = Maps.newConcurrentMap();
+  private final Map<String, ContextAwareMeter> executorJobSentMeters = Maps.newConcurrentMap();
+  private boolean instrumentationEnabled;
+  MetricContext metricContext;
+
+  public DagManagerMetrics(MetricContext metricContext, boolean instrumentationEnabled) {
```
Comment (arjun4084346) [inline gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/DagManagerMetrics.java:71]: should we just have metricContext in this signature?

File: gobblin-service/src/test/java/org/apache/gobblin/service/modules/orchestration/DagManagerTest.java (hunk 2 above)
Comment (Will-Lo) [reply gobblin-service/src/test/java/org/apache/gobblin/service/modules/orchestration/DagManagerTest.java:788]: Damn today I learned 😅 I will fix that, Intellij's syntax checker flagged it for me.

Comment (arjun4084346) [reply gobblin-service/src/test/java/org/apache/gobblin/service/modules/orchestration/DagManagerTest.java:788]: Actually, it should be in uppercase SLA, then intellij recognize the 'As' sound

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/DagManager.java [hunk 4]
```diff
@@ -1132,24 +1035,39 @@ private void cleanUp() {
...
-      for (String dagId : this.dags.keySet()) {
+      // Remove dags that are finished and emit their appropriate metrics
+      for (Map.Entry<String, Dag<JobExecutionPlan>> dagIdKeyPair : this.dags.entrySet()) {
+        String dagId = dagIdKeyPair.getKey();
+        Dag<JobExecutionPlan> dag = dagIdKeyPair.getValue();
         if (!hasRunningJobs(dagId) && !this.failedDagIdsFinishRunning.contains(dagId)) {
           String status = TimingEvent.FlowTimings.FLOW_SUCCEEDED;
           if (this.failedDagIdsFinishAllPossible.contains(dagId)) {
-            onFlowFailure(dagId);
+            if (TimingEvent.FlowTimings.FLOW_RUN_DEADLINE_EXCEEDED.equals(dag.getFlowEvent())) {
```
Comment (arjun4084346) [inline gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/DagManager.java:1059]: Are we sure we want to decouple `addFailedDag` and metrics emission?

Comment (arjun4084346) [reply gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/DagManager.java:1059]: Maybe, we can move this if block inside `addFailedDag` ?

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/DagManager.java [hunk 5]
```diff
@@ -1132,24 +1035,39 @@ private void cleanUp() {
...
+              this.dagManagerMetrics.emitFlowSlaExceededMetrics(DagManagerUtils.getFlowId(dag));
+            } else if (!TimingEvent.FlowTimings.FLOW_START_DEADLINE_EXCEEDED.equals(dag.getFlowEvent())) {
+              this.dagManagerMetrics.conditionallyMarkFlowAsState(DagManagerUtils.getFlowId(this.dags.get(dagId)),
+                  DagManager.FlowState.FAILED);
+              dagManagerMetrics.emitFlowFailedMetrics(DagManagerUtils.getFlowId(this.dags.get(dagId)));
+            }
             status = TimingEvent.FlowTimings.FLOW_FAILED;
+            addFailedDag(dagId);
             this.failedDagIdsFinishAllPossible.remove(dagId);
-            conditionallyUpdateFlowGaugeExecutionState(flowGauges, DagManagerUtils.getFlowId(this.dags.get(dagId)), FlowState.FAILED);
```
Comment (arjun4084346) [inline gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/DagManager.java:1150]: This line is intentionally removed?

Comment (arjun4084346) [reply gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/DagManager.java:1150]: Why are we removing flowGauges ?

Comment (Will-Lo) [reply gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/DagManager.java:1150]: I think it was unintentional, I went back and forth on where to add the flow gauge removal and mistakenly left it out of addFailedDag(), will add it there

**Comment by arjun4084346** [comment]
+1 
LGTM. Thank you for doing the much needed refactoring of DagManager

**Review by arjun4084346 — APPROVED** [review]

**Review by ZihanLi58 — APPROVED** [review]
+1

**Pull request merged**